flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/test/util/MiniClusterResource.java

```diff
@@ -117,6 +138,7 @@ public TestEnvironment getTestEnvironment() {
 
 	@Override
 	public void before() throws Exception {
+		miniClusterResourceConfiguration = Preconditions.checkNotNull(miniClusterResourceConfigurationSupplier.get());
```

Comment:
I think we are mixing concerns by letting the `MiniClusterResource` change its behaviour across multiple tests. I think we should not pass in a configuration supplier which can lazily instantiate a configuration. Instead if you need a different `MiniClusterResource` per test, then the test should instantiate the respective `MiniClusterResource` with the proper configuration.

Reply:
The same applies to rule chains. An easy solution would be to create a `MiniClusterWithTemporaryFolderResource` which encapsulates this logic.

Reply:
That's not "easy", it doesn't even solve the issue.

The contained `TemporaryFolder` would only be available after the construction of the `MiniClusterResource`, the folder is however necessary to create the configuration of the `MiniClusterResource`. Naturally we would like both of these to be `Rules`, but that's not possible unless the configuration is generated lazily, as i propose in the PR. Due to the `RuleChain` at the time that `MiniClusterResource#before` is called we can already access the `TemporaryFolder`.

The alternative is for the `TemporaryFolder` to be setup in a `@BeforeClass` method which is then used in the for the configuration. That would totally works, but it is really unfortunate as the setting of temporary paths is rather common and required for virtually all tests that use file-based state backends.

I agree that my proposal is quite hacky, but the question we have to ask is whether we prefer more complexity in the test vs the `MiniClusterResource`, where i would prefer the latter.